src/libmongoc/tests/test-awsauth.c

```diff
@@ -0,0 +1,515 @@
+      return false;
+   }
+   if (a->expiration.set) {
+      if (mcd_time_compare (a->expiration.value.expire_at,
+                            b->expiration.value.expire_at) != 0) {
+         return false;
+      }
+   }
+   return true;
+}
+
+// _mongoc_setenv sets an environment variable. Returns false on failure.
+static bool
+_mongoc_setenv (const char *name, const char *value)
+{
+#ifdef _WIN32
+   char *envstring;
+
+   envstring = bson_strdup_printf ("%s=%s", name, value);
+   if (0 != _putenv (envstring)) {
+      return false;
+   }
+
+   return true;
+#else
+
+   if (0 != setenv (name, value, 1)) {
+      return false;
+   }
+
+   return true;
+#endif
+}
+
+// clear_env sets all AWS environment variables to empty strings.
+static void
+clear_env (void)
+{
+   ASSERT (_mongoc_setenv ("AWS_ACCESS_KEY_ID", ""));
+   ASSERT (_mongoc_setenv ("AWS_SECRET_ACCESS_KEY", ""));
+   ASSERT (_mongoc_setenv ("AWS_SESSION_TOKEN", ""));
+}
+
+// can_setenv returns true if process is able to set environment variables and
+// get them back.
+static bool
+can_setenv (void)
+{
+   if (!_mongoc_setenv ("MONGOC_TEST_CANARY", "VALUE")) {
+      return false;
+   }
+   char *got = _mongoc_getenv ("MONGOC_TEST_CANARY");
+   if (NULL == got) {
+      return false;
+   }
+   bson_free (got);
+   return true;
+}
+
+// caching_expected returns true if MONGODB-AWS authentication is expected to
+// cache credentials. Caching is expected when credentials are not passed
+// through the URI or environment variables.
+static bool
+caching_expected (const mongoc_uri_t *uri)
+{
+   if (mongoc_uri_get_username (uri)) {
+      // AWS credentials in the URI like:
+      // "mongodb://<access_key>:<secret_key>@mongodb.example.com/?authMechanism=MONGODB-AWS"
+      // are not cached.
+      return false;
+   }
+   char *got = _mongoc_getenv ("AWS_ACCESS_KEY_ID");
+   if (NULL != got) {
+      bson_free (got);
+      // AWS credentials passed in environment are not cached.
+      return false;
+   }
+   return true;
+}
+
+// do_find runs a find command. Returns false and sets `error` on error.
+static bool
+do_find (mongoc_client_t *client, bson_error_t *error)
+{
+   bson_t *filter = bson_new ();
+   mongoc_collection_t *coll =
+      mongoc_client_get_collection (client, "aws", "coll");
+   mongoc_cursor_t *cursor = mongoc_collection_find_with_opts (
+      coll, filter, NULL /* opts */, NULL /* read prefs */);
+   const bson_t *doc;
+   while (mongoc_cursor_next (cursor, &doc))
+      ;
+   bool ok = !mongoc_cursor_error (cursor, error);
+   mongoc_cursor_destroy (cursor);
+   mongoc_collection_destroy (coll);
+   bson_destroy (filter);
+   return ok;
+}
+
+static void
+test_cache (const mongoc_uri_t *uri)
+{
+   bson_error_t error;
+   _mongoc_aws_credentials_t creds;
+
+   if (!caching_expected (uri)) {
+      printf ("Caching credentials is not expected. Skipping tests expecting "
+              "credential caching.\n");
+      return;
+   }
```

Comment:
Confirming, do we want to skip tests if expected operations are not working properly (easy to miss) instead of failing loudly? Question also applies to other skip conditions in this file.

Reply:
I think "yes" for the `caching_expected` check. The [test specification](https://github.com/mongodb/specifications/blob/c06650d86f7e47ea30cb2d992942bcec6ef155f9/source/auth/tests/mongodb-aws.rst#7cached-credentials) suggests:

> To determine whether to run the cache tests, the driver can check for the absence of the AWS_ACCESS_KEY_ID environment variable and of credentials in the URI.

The `can_setenv` check appears unnecessary on the tested distro. `can_setenv` has been removed.